### PR TITLE
Ensure static TLS region is free'd last

### PR DESCRIPTION
**Pull request by sbc100**
We were using `__cxa_thread_atexit` to register the funtion for free'ing
TLS data.  And because were were running the exit handlers
`__pthread_exit_run_handlers` before `__pthread_tsd_run_dtors` it meant
that any pthread_key descructor functions would not be able to access
TLS variables.

Instead we really need to delay free'ing of TLS data until all exit
handlers and pthread_key descructors have been run.

This was causing failure of `lsan.test_stdio_locking` due to the fact
that lsan uses TLS during pthread key destruction.  Specifically
`thread_finalize` (a pthread_key desctructor function) calls
`ThreadFinish` which in turn calls `GetCurrentThread` which uses a
thread local variable to store the current thread ID.

This change should allow #14489 to be dramatically simplified because
it will allow the use of TLS variables to implement the atexit queue.

### CODE REVIEW

**Review by tlively — APPROVED** [review]
LGTM!

**Comment by sbc100** [comment]
This change depends on https://github.com/emscripten-core/emscripten/pull/15087.. any chance you could take a look at that too?

File: system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp [hunk 1]
```diff
@@ -562,15 +562,14 @@ void InitializeInterceptors() {
   LSAN_MAYBE_INTERCEPT_PTHREAD_ATFORK;
 
   LSAN_MAYBE_INTERCEPT_STRERROR;
+#endif  // !SANITIZER_FUCHSIA && !SANITIZER_EMSCRIPTEN
```
Comment (kripken) [inline system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:565]: is the movement of the fuscia ifdef here a divergence from upstream?

Comment (kripken) [reply system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:565]: (otherwise lgtm)

Comment (sbc100) [reply system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:565]: Oh yes, I hadn't realized I had effect fuchsia here.  Will follow up.

**Pull request merged**